protocols/request-response/src/cbor.rs

```diff
@@ -44,11 +45,18 @@ pub type Behaviour<Req, Resp> = crate::Behaviour<codec::Codec<Req, Resp>>;
 
 mod codec {
     use async_trait::async_trait;
+    use cbor4ii::{core::error as error_cbor, serde as serde_cbor};
     use futures::prelude::*;
     use futures::{AsyncRead, AsyncWrite};
     use libp2p_swarm::StreamProtocol;
     use serde::{de::DeserializeOwned, Serialize};
-    use std::{io, marker::PhantomData};
+    use std::{collections::TryReserveError, convert::Infallible, io, marker::PhantomData};
+
+    #[derive(Debug)]
+    enum CborError<E> {
+        Decode(serde_cbor::DecodeError<E>),
+        Encode(serde_cbor::EncodeError<E>),
+    }
```

Comment:
Do we need this necessarily? Wouldn't it be easier to understand if we have two functions, one for `DecodeError` and one for `EncodeError` to map to an `io::Error`?

Reply:
@thomaseizinger I am re-mapping them to io::Error, but wanted to use the conversion trait which is a better abstraction, but, of course, you can't implement with types you don't know. 

This could be a function that matches over both error types from the lib, but I thought to expose the conversion trait. I can go either way on this.

Reply:
> which is a better abstraction

Not necessarily because trait implementations are part of the public API and thus can leak a dependency which is otherwise hidden and result in breaking changes.

My intuition is that two mapping functions that we call explicitly using `map_err` results in overall less code so I'd prefer that.

Reply:
@thomaseizinger ok, no worries, will work these into two functions.

Reply:
Actually, went with a private trait impl.

Reply:
Reworked this into a private trait impl (not exposed elsewhere).